_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+    }))
+```
+
+If you use `--cpu` and `crosstool_top` to define what `:android` means, what
+happens when an app supports a new Android phone with a different CPU?
+`:android` won't trigger, the app won't get required Android support libraries,
+and it will crash.
+
+### A Better Way
+
+Configurability's platform work, led by [@katre](https://github.com/katre),
+lets you declare exactly what you want in a way all rules understand. It
+doesn't even matter if no one's ever heard of your platform.
+
+With the new API, our example can be rewritten as:
+
+```python
+$ cat platforms/BUILD
+constraint_setting(name = "os")
+constraint_value(name = "android", constraint_setting = ":os")
+constraint_value(name = "linux", constraint_setting = ":os")
+
+constraint_setting(name = "device") constraint_value(name = "phone",
+constraint_setting = ":device") constraint_value(name = "tablet",
+constraint_setting = ":device")
+
+platform(name = "pixel3", constraint_values = [":android", ":phone"])
+```
+
+```python
+$ cat helpers/BUILD
+config_setting(name = "android", constraint_values = "//platforms:android")
+java_library(
+    name = "supporting_library",
+    deps = select({
+	    ":android": [":extra_required_android_deps"],
+        "//conditions:default": []
+    }))
+```
+
+and built with:
+
+    $ bazel build //my:binary --platforms=//platforms:pixel3 --define MYFEATURE=1
+
+This is more concise and correct than before. Now it doesn't matter what the
+CPU is. Any platform with the `:android`
+[`constraint_value`](https://docs.bazel.build/versions/master/be/platform.html#constraint_value)
+is an Android platform. So it's easy to express *exactly* what you want and
+have `select`, rules, and toolchains understand it the same way. `MYFEATURE=1`,
+which isn't a platform property, remains as before.
+
+Rules understand `:android` by declaring [toolchain types](https://docs.bazel.build/versions/master/toolchains.html#writing-rules-that-use-toolchains).
+For example, `java_binary` can be defined in Starlark as
+
+```python
+java_binary = rule(..., toolchains = ["//toolchain_types:java"])
+```
+
+This tells Bazel that `java_binary` understands toolchains that set
+`["//toolchain_types:java"]`. You can then write a Java toolchain for Android
+as:
+
+```python
+toolchain(
+    name = "android_jdk",
+    toolchain_type = ["//toolchain_types:java"],
+    target_compatible_with = ["//platforms:android"],
+    toolchain = ":android_jdk_provider_rule")
+```
+
+where `android_jdk_provider_rule` is a Starlark rule that
+[provides](ttps://docs.bazel.build/versions/master/skylark/rules.html#providers)
+ access to the actual JDK tools.
+
+This takes some infrastructure to set up, but the result is magic. Call your
+build with `--platforms=//platforms:pixel3` and `java_binary` automatically
+uses `:android_jdk`. All rules can join in on this. The only obligations rule
+designers have are to write Starlark describing how their toolchains work and
```

Comment:
"write Starlark" what? I think this needs "rules" or "providers" or something after Starlark.

Reply:
Done.